### PR TITLE
(CDAP-3008) Use yarn-cluster mode to run Spark program

### PR DESCRIPTION
**Pull request by chtyim**
- Added SparkExecutionService and SparkExecutionClient for handling
  communication between the driver (runs in AM) and the client.
- Added support for setting driver resources through the SparkClientContext
  in the beforeSubmit()

### CODE REVIEW

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/distributed/ContainerLauncherGenerator.java [hunk 1]
```diff
@@ -133,6 +148,55 @@ private static void generateLauncherClass(String launcherClassPath, String class
     output.write(classWriter.toByteArray());
   }
 
+  /**
+   * Generates a class that has a static main method which delegates the call to another main method.
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/distributed/ContainerLauncherGenerator.java]: typo? This actually seems to be calling the ``launch`` method?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/distributed/ContainerLauncherGenerator.java [hunk 2]
```diff
@@ -133,6 +148,55 @@ private static void generateLauncherClass(String launcherClassPath, String class
...
+    GeneratorAdapter mg = new GeneratorAdapter(Opcodes.ACC_PUBLIC, constructor, null, null, classWriter);
+    mg.loadThis();
+    mg.invokeConstructor(Type.getType(Object.class), constructor);
+    mg.returnValue();
+    mg.endMethod();
+
+    // Generate the main method
+    // public static void main(String[] args) {
+    //   System.out.println("Launch class .....");
+    //   <MainDelegator>.main(<className>, args);
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/distributed/ContainerLauncherGenerator.java]: ditto?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedWorkflowProgramRunner.java [hunk 3]
```diff
@@ -134,11 +134,11 @@ private static YarnConfiguration createConfiguration(YarnConfiguration hConf) {
...
   private DriverMeta findDriverResources(Map<String, SparkSpecification> sparkSpecs,
-                                        Map<String, MapReduceSpecification> mrSpecs,
-                                        WorkflowSpecification spec) {
-    // Find the resource requirements from the workflow
+                                         Map<String, MapReduceSpecification> mrSpecs,
+                                         WorkflowSpecification spec) {
+    // Find the resource requirements from the workflow with 768MB as minimum.
     // It is the largest memory and cores from all Spark and MapReduce programs inside the workflow
-    Resources resources = new Resources();
+    Resources resources = new Resources(768);
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedWorkflowProgramRunner.java]: why 768?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractDistributedProgramRunner.java [hunk 4]
```diff
@@ -186,6 +186,12 @@ public final ProgramController run(final Program program, final ProgramOptions o
         public TwillController launch(TwillApplication twillApplication, Iterable<String> extraClassPaths,
                                       Iterable<? extends Class<?>> extraDependencies) {
           TwillPreparer twillPreparer = twillRunner.prepare(twillApplication);
+          // TODO: CDAP-5506. It's a bit hacky to set a Spark environment here. However, we always launch
+          // Spark using YARN and it is needed for both Workflow and Spark runner. We need to set it
+          // because inside Spark code, it will set and unset the SPARK_YARN_MODE system properties, causing
+          // fork in distributed mode not work. By setting it in the environment, which Spark always use it
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/AbstractDistributedProgramRunner.java]: nit: Setting it in the environment, which Spark uses for defaults, so it can't be unset by Spark

File: cdap-common/src/main/java/co/cask/cdap/common/lang/ClassLoaders.java [hunk 5]
```diff
@@ -79,4 +83,20 @@ public static ClassLoader setContextClassLoader(ClassLoader classLoader) {
     // The casting should succeed since it's either null or is assignable to the given type
     return (T) result;
   }
+
+  /**
+   * Populates the list of {@link URL} that this ClassLoader uses, including all URLs used by the parent of the
+   * given ClassLoader.
+   */
+  public static List<URL> getClassLoaderURLs(ClassLoader classLoader, List<URL> urls) {
```
Comment (bdmogal) [inline cdap-common/src/main/java/co/cask/cdap/common/lang/ClassLoaders.java]: ``classLoader`` is ``@Nullable``

Comment (bdmogal) [inline cdap-common/src/main/java/co/cask/cdap/common/lang/ClassLoaders.java]: also, we seem to add to the passed ``urls`` parameter, right? So should we call this ``addClassLoaderURLs``?

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeContextConfig.java [hunk 6]
```diff
@@ -139,7 +139,6 @@ public WorkflowProgramInfo getWorkflowProgramInfo() {
       return null;
     }
     WorkflowProgramInfo workflowProgramInfo = GSON.fromJson(info, WorkflowProgramInfo.class);
-    workflowProgramInfo.getWorkflowToken().disablePut();
```
Comment (bdmogal) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeContextConfig.java]: unrelated change?

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeService.java [hunk 7]
```diff
@@ -168,9 +171,11 @@ public SparkExecutionContext create(SparkRuntimeContext runtimeContext) {
         File expandedProgramJar = Locations.linkOrCopy(runtimeContext.getProgram().getJarLocation(),
                                                        new File(tempDir,
                                                                 SparkRuntimeContextProvider.PROGRAM_JAR_EXPANDED_NAME));
-        // Localize both the unexpanded and expanded jar
-        localizeResources.add(new LocalizeResource(programJar, false));
+        // Localize both the unexpanded and expanded jar program jar
```
Comment (bdmogal) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeService.java]: typo: extra 'jar'

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeService.java [hunk 8]
```diff
@@ -168,9 +171,11 @@ public SparkExecutionContext create(SparkRuntimeContext runtimeContext) {
...
+        localizeResources.add(new LocalizeResource(programJar));
```
Comment (bdmogal) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeService.java]: this is an archive right? Shouldn't it be ``new LocalizeResource(programJar, true)``?

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedWorkflowProgramRunner.java (hunk 3 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedWorkflowProgramRunner.java]: The `Resources()` has default of 512, which is easily not enough for the Workflow driver. Setting to 768 is temporary. There is work to improve resources setting for Workflow (CDAP-5066).

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeService.java [hunk 9]
```diff
@@ -362,15 +386,16 @@ public Void call() throws Exception {
       }
     }
 
-    String extraClassPath = "$PWD/" + CDAP_SPARK_JAR + "/lib/*";
+    String extraClassPath = "$PWD/" + CDAP_LAUNCHER_JAR + File.pathSeparator + "$PWD/" + CDAP_SPARK_JAR + "/lib/*";
```
Comment (bdmogal) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeService.java]: is ``CDAP_LAUNCHER_JAR`` a directory? Also, should we use ``Joiner.on(File.pathSeparator)`` here instead of using the forward slashes

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeService.java (hunk 8 above)
Comment (chtyim) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeService.java]: No. This is the jar, the`"expandedProgramJar` is the archive

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeService.java (hunk 9 above)
Comment (chtyim) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeService.java]: CDAP_LAUNCHER_JAR is a jar. Thought about using Joiner, but doesn't seems give better code. There are two separators involves. The `pathSeparator == :` and the `separator == /`.

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeContextConfig.java (hunk 6 above)
Comment (chtyim) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeContextConfig.java]: This change is necessary. When the driver is running in the Spark AM, the workflow token is deserialized through this method. The driver can modify the token, hence we cannot disable put on the token.

File: cdap-common/src/main/java/co/cask/cdap/common/lang/ClassLoaders.java (hunk 5 above)
Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/common/lang/ClassLoaders.java]: Hum, the `ClassLoader` is not intended to be `nullable`. Will remove the check.

Comment (chtyim) [inline cdap-common/src/main/java/co/cask/cdap/common/lang/ClassLoaders.java]: I think `get` is fine since it is getting the `URL`s from the given ClassLoader. The collection is for storing the result.

**Comment by anew** [comment]
Can you explain why this is needed? For MapReduce, we run the MR client in the same JVM as the program runner. This will add one more container to every Spark job? That is, two containers for the Twill app (AM and submitter) plus the driver, plus all the executors.

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkContainerLauncher.java [hunk 10]
```diff
@@ -0,0 +1,88 @@
...
+    }
+
+    if (!urls.remove(getClassPathURL(mainClassName, resource))) {
+      throw new IllegalStateException("Failed to remove main class resource " + resource);
+    }
+
+    ClassLoader classLoader = new SparkRunnerClassLoader(urls.toArray(new URL[urls.size()]),
+                                                         ClassLoader.getSystemClassLoader().getParent(),
+                                                         false);
+    Thread.currentThread().setContextClassLoader(classLoader);
```
Comment (bdmogal) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkContainerLauncher.java]: we never have to reset this?

Comment (bdmogal) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkContainerLauncher.java]: Probably because this will just launch the spark job and exit? Perhaps a comment here would help?

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkContainerLauncher.java [hunk 11]
```diff
@@ -0,0 +1,88 @@
...
+    if (resource == null) {
+      throw new IllegalStateException("Failed to find resource for main class " + mainClassName);
+    }
+
+    if (!urls.remove(getClassPathURL(mainClassName, resource))) {
+      throw new IllegalStateException("Failed to remove main class resource " + resource);
+    }
+
+    ClassLoader classLoader = new SparkRunnerClassLoader(urls.toArray(new URL[urls.size()]),
+                                                         ClassLoader.getSystemClassLoader().getParent(),
```
Comment (bdmogal) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkContainerLauncher.java]: should we document why this is the parent?

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkContainerLauncher.java [hunk 12]
```diff
@@ -0,0 +1,88 @@
...
+   * @param args arguments for the main class
+   */
+  @SuppressWarnings("unused")
+  public static void launch(String mainClassName, String[] args) {
+    List<URL> urls = ClassLoaders.getClassLoaderURLs(ClassLoader.getSystemClassLoader(), new ArrayList<URL>());
+
+    // Remove the URL that contains the given main classname to avoid infinite recursion.
+    // This is needed because we generate a class with the same main classname in order to intercept the main()
+    // method call from the container launch script.
+    URL resource = ClassLoader.getSystemClassLoader().getResource(mainClassName.replace('.', '/') + ".class");
```
Comment (bdmogal) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkContainerLauncher.java]: we have this pattern: ``mainClassName.replace('.', '/') + ".class"`` at multiple places right? Is it worth creating a helper method for this?

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkExecutionClient.java [hunk 13]
```diff
@@ -0,0 +1,135 @@
...
+   *
+   * @param workflowToken the {@link BasicWorkflowToken} to send in the heartbeat request if it is not {@code null}.
+   *                      only changes made by this Spark program will be sent.
+   * @return the {@link SparkCommand} returned by the service or {@code null} if there is no command.
+   */
+  @Nullable
+  public SparkCommand heartbeat(@Nullable BasicWorkflowToken workflowToken) throws Exception {
+    HttpURLConnection urlConn = openConnection("heartbeat");
+    try {
+      urlConn.setRequestMethod("POST");
```
Comment (bdmogal) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkExecutionClient.java]: should we use a constant here?

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkExecutionClient.java [hunk 14]
```diff
@@ -0,0 +1,135 @@
...
+  /**
+   * Sends a execution completed request to the execution service.
+   *
+   * @param workflowToken the {@link BasicWorkflowToken} to send in the heartbeat request if it is not {@code null}.
+   *                      only changes made by this Spark program will be sent.
+   */
+  public void completed(@Nullable BasicWorkflowToken workflowToken) throws Exception {
+    HttpURLConnection urlConn = openConnection("completed");
+    try {
+      urlConn.setRequestMethod("PUT");
```
Comment (bdmogal) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkExecutionClient.java]: ditto

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkExecutionService.java [hunk 15]
```diff
@@ -0,0 +1,205 @@
...
+        );
+      }
+    }
+
+    /**
+     * Updates {@link WorkflowToken} of the program. It is a json Map<String, String> in the request body.
+     */
+    private void updateWorkflowToken(ChannelBuffer requestBody) {
+      if (!requestBody.readable()) {
+        return;
```
Comment (bdmogal) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkExecutionService.java]: should we log a debug/trace here?

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkExecutionClient.java (hunk 13 above)
Comment (chtyim) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkExecutionClient.java]: Constant for "POST" and is being used one time?

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkExecutionService.java (hunk 15 above)
Comment (chtyim) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkExecutionService.java]: Why? It is part of the protocol and is expected. See https://issues.cask.co/browse/CDAP-3008 for the protocol

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkExecutionClient.java (hunk 14 above)
Comment (chtyim) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkExecutionClient.java]: I would say no.

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkContainerLauncher.java (hunk 12 above)
Comment (chtyim) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkContainerLauncher.java]: Not sure. I don't like one line helper method flying around.

**Comment by chtyim** [comment]
@anew the number of containers are the same for yarn-client and yarn-cluster modes. It is also the same as in MR case. Basically two for the client (Twill AM + Twill Runnable) and then N+1 for the Spark execution (1 Spark AM + N Spark executors).

Using yarn-cluster mode gives better resources isolation and preventing drivers affecting each other. It also works better for failure recovery (YARN will restart the Spark job if the AM fails up to max attempts). Moreover, the Spark UI, which runs in the driver, uses system properties to set the UI host:port and pass it to the Yarn AM when running in yarn-client mode and it won't work when there is a fork. By isolating the driver using yarn-cluster mode, we don't suffer from this problem.

File: cdap-spark-core/src/main/scala/co/cask/cdap/app/runtime/spark/SparkMainWrapper.scala [hunk 16]
```diff
@@ -17,53 +17,88 @@
...
+          // JavaSparkMain
+          case cls if classOf[JavaSparkMain].isAssignableFrom(cls) =>
+            cls.asSubclass(classOf[JavaSparkMain]).newInstance().run(
+              sparkClassLoader.createJavaExecutionContext(executionContext))
+
+          // main() method
+          case cls =>
+            getMainMethod(cls).fold(
+              throw new IllegalArgumentException(userSparkClass.getName
+                + " is not a supported Spark program. It should implements either "
```
Comment (bdmogal) [inline cdap-spark-core/src/main/scala/co/cask/cdap/app/runtime/spark/SparkMainWrapper.scala]: nit: implement

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedWorkflowProgramRunner.java (hunk 3 above)
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedWorkflowProgramRunner.java]: should we add this info as a comment?

Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/distributed/DistributedWorkflowProgramRunner.java]: I'll just leave it to CDAP-5066 since it will change this method anyway.

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeService.java (hunk 9 above)
Comment (bdmogal) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeService.java]: Oh I see. However, this won't be safe on non-unix filesystems, right?

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkExecutionClient.java (hunk 13 above)
Comment (bdmogal) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkExecutionClient.java]: I didn't mean constant in this class. Something like ``HttpMethod.POST``?

Comment (chtyim) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkExecutionClient.java]: No. It doesn't give more readable code. It will become `urlConn.setRequestMethod(HttpMethod.POST.name())`.

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeService.java (hunk 9 above)
Comment (chtyim) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeService.java]: Probably not. Can change that to use `File.separate` to construct the path. Finding if there is better library for doing the job.

**Comment by anew** [comment]
@chtyim, I understand the reason for yarn cluster mode now. Also I was mistaken about the number of containers. +1

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/distributed/ContainerLauncherGenerator.java (hunk 1 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/distributed/ContainerLauncherGenerator.java]: fixed

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/distributed/ContainerLauncherGenerator.java (hunk 2 above)
Comment (chtyim) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/batch/distributed/ContainerLauncherGenerator.java]: fixed.

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeService.java (hunk 7 above)
Comment (chtyim) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeService.java]: fixed

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeService.java (hunk 9 above)
Comment (chtyim) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRuntimeService.java]: Changed to use `Paths` and `Joiner`.

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkContainerLauncher.java (hunk 10 above)
Comment (chtyim) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkContainerLauncher.java]: Added comments.

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkContainerLauncher.java (hunk 11 above)
Comment (chtyim) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/distributed/SparkContainerLauncher.java]: Added comment.

**Comment by bdmogal** [comment]
lgtm

File: cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactClassLoaderFactory.java [hunk 17]
```diff
@@ -73,8 +73,14 @@ CloseableClassLoader createClassLoader(Location artifactLocation) throws IOExcep
...
       // TODO: CDAP-5613. We shouldn't always expose the Spark classes.
       ProgramRunner programRunner = programRunnerFactory.create(ProgramType.SPARK);
-      if (programRunner instanceof ProgramClassLoaderProvider) {
-        programClassLoader = ((ProgramClassLoaderProvider) programRunner).createProgramClassLoader(cConf, unpackDir);
+      try {
+        if (programRunner instanceof ProgramClassLoaderProvider) {
+          programClassLoader = ((ProgramClassLoaderProvider) programRunner).createProgramClassLoader(cConf, unpackDir);
+        }
+      } finally {
+        if (programRunner instanceof Closeable) {
```
Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactClassLoaderFactory.java]: which ``ProgramRunner`` classes are ``Closeable``

Comment (bdmogal) [inline cdap-app-fabric/src/main/java/co/cask/cdap/internal/app/runtime/artifact/ArtifactClassLoaderFactory.java]: Oh nvm. I see that the ``SparkProgramRunner`` is made ``Closeable`` later in this commit. I was looking at the release branch and couldn't find it.

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkProgramRunner.java [hunk 18]
```diff
@@ -202,33 +202,27 @@ public ProgramClassLoader createProgramClassLoader(CConfiguration cConf, File di
...
-              LOG.debug("Closed ClassLoader for SparkProgramRunner of {}, run {}.", programId, runId);
-            }
-          }
-        };
-        t.setDaemon(true);
-        t.start();
+      public void run() {
+        // Delay the closing of the ClassLoader because Spark, which uses akka, has an async cleanup process
+        // for shutting down threads. During shutdown, there are new classes being loaded.
+        Uninterruptibles.sleepUninterruptibly(2, TimeUnit.SECONDS);
```
Comment (bdmogal) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkProgramRunner.java]: no need to make this configurable, right? Is 2 seconds sufficient always - guessing there is no reliable way of knowing when that async cleanup process is completed (listener or anything like that)?

**Comment by bdmogal** [comment]
Just 1 question in the last commit. Rest LGTM

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkProgramRunner.java (hunk 18 above)
Comment (chtyim) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkProgramRunner.java]: No need to be configurable. In fact, if the classloader get closed before the akka system do shutdown, it won't affect anything, except a stupid error message from akka.

**Comment by bdmogal** [comment]
LGTM

**Comment by chtyim** [comment]
Build passed: http://builds.cask.co/browse/CDAP-DUT3910-13

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRunnerClassLoader.java [hunk 19]
```diff
@@ -131,9 +131,13 @@ public SparkRunnerClassLoader(URL[] urls, @Nullable ClassLoader parent, boolean
     // cdap-api-classes
     // Any class that is not from cdap-api-spark or cdap-spark-core
     // Any class that is not from Spark
+    // We also need to define the org.spark-project., fastxml and akka classes in this ClassLoader
+    // to avoid reference/thread leakage due to Spark assumption on process terminating after execution
     if (API_CLASSES.contains(name) || (!name.startsWith("co.cask.cdap.api.spark.")
```
Comment (bdmogal) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRunnerClassLoader.java]: when/how is this classloader cleaned up?

**Comment by bdmogal** [comment]
1 question. Rest LGTM

File: cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRunnerClassLoader.java (hunk 19 above)
Comment (chtyim) [inline cdap-spark-core/src/main/java/co/cask/cdap/app/runtime/spark/SparkRunnerClassLoader.java]: It's done by the caller. See https://github.com/caskdata/cdap/pull/5591/commits/e02886cc39bf4660b2790dac193d20b5d0d78c3e
